basic_context.go

```diff
@@ -1,21 +1,51 @@
+// Copyright (C) 2022 Storj Labs, Inc.
+// See LICENSE for copying information.
+
 package clingy
 
 import (
+	"bytes"
 	"context"
 	"io"
+	"os"
 )
 
-type basicContext struct {
-	context.Context
+type stdioContextKey string
+
+const stdioKey stdioContextKey = "environment"
 
+type stdioEnvironment struct {
 	stdin  io.Reader
 	stdout io.Writer
 	stderr io.Writer
 }
 
-func (b basicContext) Read(p []byte) (n int, err error)  { return b.stdin.Read(p) }
-func (b basicContext) Write(p []byte) (n int, err error) { return b.stdout.Write(p) }
+// WrapStdio saves stdin/out/err to the context for later use.
+func WrapStdio(ctx context.Context) context.Context {
+	return context.WithValue(ctx, stdioKey, stdioEnvironment{
+		stdout: os.Stdout,
+		stdin:  os.Stdin,
+		stderr: os.Stderr,
+	})
+}
+
+// WithBufferedStdio creates context with in-memory stdio targets.
+func WithBufferedStdio(ctx context.Context) context.Context {
+	return context.WithValue(ctx, stdioKey, stdioEnvironment{
+		stdout: &bytes.Buffer{},
+		stdin:  &bytes.Buffer{},
+		stderr: &bytes.Buffer{},
+	})
+}
+
+func Stdin(ctx context.Context) io.Reader {
+	return ctx.Value(stdioKey).(stdioEnvironment).stdin
```

Comment:
This will panic if called on a context that has not been wrapped. This is why I previously had it written as

```
func Stdin(ctx context.Context) io.Reader {
    env, _ := ctx.Value(stdioKey).(stdioEnvironment)
    return env.stdin
}```
The `_` part is to avoid the panic.